pkg/authorization/api/conversion_test.go

```diff
@@ -0,0 +1,352 @@
+	for i := range subjects {
+		subject := &subjects[i]
+		subject.APIVersion = rbac.GroupName
+		setValidRBACKindAndNamespace(subject, i, c)
+	}
+	roleRef.APIGroup = rbac.GroupName
+	roleRef.Kind = getKind(namespace)
+}
+
+func setValidRBACKindAndNamespace(subject *rbac.Subject, i int, c fuzz.Continue) {
+	kinds := []string{rbac.UserKind, rbac.GroupKind, rbac.ServiceAccountKind}
+	kind := kinds[c.Intn(len(kinds))]
+	subject.Kind = kind
+
+	if subject.Kind != rbac.ServiceAccountKind {
+		subject.Namespace = ""
+	}
+
+	switch subject.Kind {
+
+	case rbac.UserKind:
+		if len(uservalidation.ValidateUserName(subject.Name, false)) != 0 {
+			subject.Name = fmt.Sprintf("validusername%d", i)
+		}
+
+	case rbac.GroupKind:
+		if len(uservalidation.ValidateGroupName(subject.Name, false)) != 0 {
+			subject.Name = fmt.Sprintf("validgroupname%d", i)
+		}
+
+	case rbac.ServiceAccountKind:
+		if len(validation.ValidateNamespaceName(subject.Namespace, false)) != 0 {
+			subject.Namespace = fmt.Sprintf("sanamespacehere%d", i)
+		}
+		if len(validation.ValidateServiceAccountName(subject.Name, false)) != 0 {
+			subject.Name = fmt.Sprintf("sanamehere%d", i)
+		}
+
+	default:
+		panic("invalid kind")
+	}
+}
+
+func setRandomOriginRoleBindingData(subjects []api.ObjectReference, roleRef *api.ObjectReference, namespace string, c fuzz.Continue) {
+	for i := range subjects {
+		subject := &subjects[i]
+		unsetUnpreservedOriginFields(subject)
+		setValidOriginKindAndNamespace(subject, i, c)
+	}
+	unsetUnpreservedOriginFields(roleRef)
+	roleRef.Kind = getKind(namespace)
+	roleRef.Namespace = namespace
+}
+
+func setValidOriginKindAndNamespace(subject *api.ObjectReference, i int, c fuzz.Continue) {
+	kinds := []string{UserKind, SystemUserKind, GroupKind, SystemGroupKind, ServiceAccountKind}
+	kind := kinds[c.Intn(len(kinds))]
+	subject.Kind = kind
+
+	if subject.Kind != ServiceAccountKind {
+		subject.Namespace = ""
+	}
+
+	switch subject.Kind {
+
+	case UserKind:
+		if len(uservalidation.ValidateUserName(subject.Name, false)) != 0 {
+			subject.Name = fmt.Sprintf("validusername%d", i)
+		}
+
+	case GroupKind:
+		if len(uservalidation.ValidateGroupName(subject.Name, false)) != 0 {
+			subject.Name = fmt.Sprintf("validgroupname%d", i)
+		}
+
+	case SystemUserKind, SystemGroupKind:
+		subject.Name = ":" + subject.Name
+
+	case ServiceAccountKind:
+		if len(validation.ValidateNamespaceName(subject.Namespace, false)) != 0 {
+			subject.Namespace = fmt.Sprintf("sanamespacehere%d", i)
+		}
+		if len(validation.ValidateServiceAccountName(subject.Name, false)) != 0 {
+			subject.Name = fmt.Sprintf("sanamehere%d", i)
+		}
+
+	default:
+		panic("invalid kind")
+	}
+}
+
+func unsetUnpreservedOriginFields(ref *api.ObjectReference) {
+	ref.UID = ""
+	ref.ResourceVersion = ""
+	ref.FieldPath = ""
+	ref.APIVersion = ""
+}
+
+func sortAndDeduplicateRBACRulesFields(in []rbac.PolicyRule) {
```

Comment:
Is this a fancy reflective mess for creating sets and calling .List on a half dozen fields?  Do that instead.